Source/CesiumRuntime/Private/Cesium3DTileset.cpp

```diff
@@ -1581,6 +1646,142 @@ void ACesium3DTileset::showTilesToRender(
   }
 }
 
+namespace {
+
+template <typename Tag, typename Tag::type M> struct Rob {
+public:
+  friend typename Tag::type get(Tag) { return M; }
+};
+
+// Hack to retrieve the private ULocalPlayer::ViewStates
+class ULocalPlayer_Hack {
+public:
+  typedef TArray<FSceneViewStateReference> ULocalPlayer::*type;
+  friend type get(ULocalPlayer_Hack);
+};
+
+template struct Rob<ULocalPlayer_Hack, &ULocalPlayer::ViewStates>;
+} // namespace
+
+void ACesium3DTileset::RetrieveOcclusionResults(
+    TArray<FSceneViewState*>&& views) {
+  this->OcclusionResults.clear();
+  this->OcclusionResults.reserve(this->OcclusionRequests.size());
+  this->WaitingForOcclusion = true;
+  ENQUEUE_RENDER_COMMAND(RetrieveOccludedPrimitives)
+  ([views = std::move(views),
+    &WaitingForOcclusion = this->WaitingForOcclusion,
+    &OcclusionRequests = this->OcclusionRequests,
+    &OcclusionResults = this->OcclusionResults](FRHICommandList& RHICmdList) {
+    for (const OcclusionRequest& occlusionRequest : OcclusionRequests) {
+      // Check that the primitive is definitely occluded in every view.
+      bool isOccluded = false;
+      bool isDefinite = true;
+      for (FSceneViewState* pViewState : views) {
+        if (pViewState && pViewState->PrimitiveOcclusionHistorySet.Num()) {
+          FPrimitiveOcclusionHistory* pHistory =
+              pViewState->PrimitiveOcclusionHistorySet.Find(
+                  FPrimitiveOcclusionHistoryKey(
+                      occlusionRequest.primitiveId,
+                      0));
+          if (pHistory) {
+            if (!pHistory->OcclusionStateWasDefiniteLastFrame) {
+              isDefinite = false;
+              break;
+            }
+
+            isDefinite = true;
+
+            if (occlusionRequest.currentOcclusionState) {
+              if (pHistory->LastPixelsPercentage > 0.01f) {
+                isOccluded = false;
+                break;
+              }
+            } else {
+              if (!pHistory->WasOccludedLastFrame) {
+                // pHistory->LastPixelsPercentage > 0.0f) {
+                isOccluded = false;
+                break;
+              }
+            }
+
+            isOccluded = true;
+          }
+        }
+      }
+
+      if (isDefinite) {
+        OcclusionResults.emplace(
+            occlusionRequest.primitiveId.PrimIDValue,
+            isOccluded);
+      }
+    }
+
+    WaitingForOcclusion = false;
+  });
+}
+
+void ACesium3DTileset::TickOcclusionHandling() {
+  if (!this->WaitingForOcclusion) {
+    TArray<UCesiumBoundingVolumeComponent*> boundingVolumes;
+    this->GetComponents<UCesiumBoundingVolumeComponent>(boundingVolumes, false);
+
+    this->OcclusionRequests.clear();
+    this->OcclusionRequests.reserve(boundingVolumes.Num());
+
+    // We are not waiting on the render thread, we can let the primitives know
+    // about the last-received occlusion info.
+    for (UCesiumBoundingVolumeComponent* pBoundingVolume : boundingVolumes) {
```

Comment:
We no longer have to do the allocation and pointer copies in the current state of the branch. Another simplification is that only the UCesiumBoundingVolumePoolComponent needs to be found by class. The actual bounding volume components will all be attached to the pool component so we just iterate through all the attached children of the pool component. 

One thing that could still use improvement is the fact that we currently search for the pool + iterate through all the BVs once per each view + once to finalize the multi-view occlusion aggregation each frame. I will profile to see if it is a hotspot, maybe we should just save a pointer to the UBoundingVolumePoolComponent, this may introduce some complication with deletion and GC reference counting that need to be sorted out though.